pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java

```diff
@@ -496,30 +496,42 @@ public CompletableFuture<Optional<BrokerLookupData>> assign(Optional<ServiceUnit
             if (topic.isPresent() && isInternalTopic(topic.get().toString())) {
                 owner = serviceUnitStateChannel.getChannelOwnerAsync();
             } else {
-                String candidateBrokerId = getHeartbeatOrSLAMonitorBrokerId(serviceUnit);
-                if (candidateBrokerId != null) {
-                    owner = CompletableFuture.completedFuture(Optional.of(candidateBrokerId));
-                } else {
-                    owner = getOrSelectOwnerAsync(serviceUnit, bundle).thenApply(Optional::ofNullable);
-                }
+                owner = getHeartbeatOrSLAMonitorBrokerId(serviceUnit).thenCompose(candidateBrokerId -> {
+                    if (candidateBrokerId != null) {
+                        return CompletableFuture.completedFuture(Optional.of(candidateBrokerId));
+                    }
+                    return getOrSelectOwnerAsync(serviceUnit, bundle).thenApply(Optional::ofNullable);
+                });
             }
             return getBrokerLookupData(owner, bundle);
         });
     }
 
-    private String getHeartbeatOrSLAMonitorBrokerId(ServiceUnitId serviceUnit) {
+    private CompletableFuture<String> getHeartbeatOrSLAMonitorBrokerId(ServiceUnitId serviceUnit) {
```

Comment:
Could we add a public method to `NamespaceService` and reuse that method so that we can avoid duplicated logic here? I think `NamespaceService#isBrokerActive` can be replaced by the similar logic because it calls `LoadManager#getAvailableBrokers`.

Reply:
We can add a `LockManager<BrokerLookupData>` field to `NamespaceService` and apply the similar logic of `BrokerRegistryImpl#lookupAsync` to check if the broker is alive.

```java
 this.brokerLookupDataLockManager = pulsar.getCoordinationService().getLockManager(BrokerLookupData.class);
```

Reply:
For example,

```java
    public static CompletableFuture<String> getHeartbeatOrSLAMonitorBrokerId(
            ServiceUnitId serviceUnit, LockManager<BrokerLookupData> brokerLookupData) {
```

Reply:
This change requires making the `NamespaceService#searchForCandidateBroker` async and introducing many changes.

Reply:
You can just keep the synchronous way by calling `get` with the configured metadata timeout.